File: src/dropbox.rs
use anyhow::{Context, Result};
use base64::{Engine as _, engine::general_purpose::STANDARD as BASE64};
use ini::Ini;
use std::fs;

/// The official Dropbox client stores a base64-encoded path in host.db.
const DROPBOX_HOST_DB: &str = ".dropbox/host.db";
/// Maestral stores its sync folder in an INI file under the home directory.
const MAESTRAL_INI: &str = "Library/Application Support/maestral/maestral.ini";

#[derive(Default)]
pub struct DropBox {
    pub path: String,
}

impl DropBox {
    pub fn new() -> Self {
        Self::default()
    }

    /// Resolve and cache the Dropbox folder. Returns an empty string when no
    /// Dropbox or Maestral installation is found.
    pub fn folder(&mut self) -> Result<&str> {
        if self.path.is_empty() {
            self.path = Self::resolve_folder()?;
        }
        Ok(&self.path)
    }

    fn resolve_folder() -> Result<String> {
        let home = dirs::home_dir().context("could not determine home directory")?;
        let host_db = home.join(DROPBOX_HOST_DB);
        let maestral = home.join(MAESTRAL_INI);

        if host_db.exists() {
            let contents = fs::read_to_string(&host_db)
                .with_context(|| format!("reading {}", host_db.display()))?;
            parse_host_db(&contents)
        } else if maestral.exists() {
            let contents = fs::read_to_string(&maestral)
                .with_context(|| format!("reading {}", maestral.display()))?;
            parse_maestral_ini(&contents)
        } else {
            Ok(String::new())
        }
    }

    pub fn name(&self) -> &str {
        self.path.rsplit('/').next().unwrap_or_default()
    }
}

/// Decode the Dropbox folder path from the contents of `.dropbox/host.db`,
/// whose second line is the base64-encoded UTF-8 path.
fn parse_host_db(contents: &str) -> Result<String> {
    let encoded = contents
        .lines()
        .nth(1)
        .context("host.db did not contain a path on its second line")?;
    let bytes = BASE64
        .decode(encoded.as_bytes())
        .context("decoding base64 path from host.db")?;
    String::from_utf8(bytes).context("Dropbox path in host.db was not valid UTF-8")
}

/// Read the sync folder path from the contents of Maestral's `maestral.ini`.
fn parse_maestral_ini(contents: &str) -> Result<String> {
    let conf = Ini::load_from_str(contents).context("parsing maestral.ini")?;
    let path = conf
        .section(Some("sync"))
        .and_then(|s| s.get("path"))
        .context("maestral.ini is missing a [sync] path entry")?;
    Ok(path.to_string())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn host_db_decodes_base64_second_line() {
        let encoded = BASE64.encode("/Users/me/Dropbox");
        let contents = format!("first-line-is-ignored\n{encoded}\n");
        assert_eq!(parse_host_db(&contents).unwrap(), "/Users/me/Dropbox");
    }

    #[test]
    fn host_db_missing_second_line_errors() {
        assert!(parse_host_db("only-one-line").is_err());
    }

    #[test]
    fn host_db_invalid_base64_errors() {
        assert!(parse_host_db("line1\nthis is not base64!!").is_err());
    }

    #[test]
    fn maestral_reads_sync_path() {
        let ini = "[sync]\npath = /Users/me/Dropbox\n";
        assert_eq!(parse_maestral_ini(ini).unwrap(), "/Users/me/Dropbox");
    }

    #[test]
    fn maestral_missing_path_errors() {
        let ini = "[main]\nfoo = bar\n";
        assert!(parse_maestral_ini(ini).is_err());
    }

    #[test]
    fn name_returns_last_path_segment() {
        let dbx = DropBox {
            path: "/Users/me/Dropbox".to_string(),
        };
        assert_eq!(dbx.name(), "Dropbox");
    }

    #[test]
    fn name_handles_no_slash_and_empty_path() {
        assert_eq!(DropBox { path: "Dropbox".into() }.name(), "Dropbox");
        assert_eq!(DropBox::new().name(), "");
    }
}

File: zig/src/dropbox.zig
const std = @import("std");
const Io = std.Io;
const Dir = std.Io.Dir;

const HOST_DB = ".dropbox/host.db";
const MAESTRAL_INI = "Library/Application Support/maestral/maestral.ini";

/// Resolve the Dropbox sync folder, or null if no Dropbox/Maestral install is
/// found. Mirrors the Rust dropbox.rs: the official client stores a
/// base64-encoded path on the 2nd line of host.db; Maestral stores it as the
/// `[sync] path` key in maestral.ini.
pub fn resolveFolder(io: Io, alloc: std.mem.Allocator, home: []const u8) !?[]const u8 {
    const cwd = Dir.cwd();

    const host_db = try std.fs.path.join(alloc, &.{ home, HOST_DB });
    if (cwd.readFileAlloc(io, host_db, alloc, std.Io.Limit.limited(64 * 1024))) |contents| {
        return parseHostDb(alloc, contents);
    } else |_| {}

    const ini = try std.fs.path.join(alloc, &.{ home, MAESTRAL_INI });
    if (cwd.readFileAlloc(io, ini, alloc, std.Io.Limit.limited(1024 * 1024))) |contents| {
        return parseMaestralIni(contents);
    } else |_| {}

    return null;
}

fn parseHostDb(alloc: std.mem.Allocator, contents: []const u8) !?[]const u8 {
    var lines = std.mem.splitScalar(u8, contents, '\n');
    _ = lines.next(); // first line is ignored
    const encoded = lines.next() orelse return null;
    const trimmed = std.mem.trim(u8, encoded, " \r\n");
    if (trimmed.len == 0) return null;

    const dec = std.base64.standard.Decoder;
    const n = dec.calcSizeForSlice(trimmed) catch return null;
    const buf = try alloc.alloc(u8, n);
    dec.decode(buf, trimmed) catch return null;
    return buf;
}

fn parseMaestralIni(contents: []const u8) ?[]const u8 {
    var in_sync = false;
    var lines = std.mem.splitScalar(u8, contents, '\n');
    while (lines.next()) |raw| {
        const line = std.mem.trim(u8, raw, " \t\r");
        if (line.len == 0 or line[0] == ';' or line[0] == '#') continue;
        if (line[0] == '[') {
            in_sync = std.mem.eql(u8, line, "[sync]");
            continue;
        }
        if (!in_sync) continue;
        const eq = std.mem.indexOfScalar(u8, line, '=') orelse continue;
        const key = std.mem.trim(u8, line[0..eq], " \t");
        if (std.mem.eql(u8, key, "path")) {
            return std.mem.trim(u8, line[eq + 1 ..], " \t");
        }
    }
    return null;
}

test "parseHostDb decodes the base64 second line" {
    const path = "/Users/me/Dropbox";
    var enc_buf: [64]u8 = undefined;
    const enc = std.base64.standard.Encoder.encode(&enc_buf, path);
    var line_buf: [128]u8 = undefined;
    const contents = try std.fmt.bufPrint(&line_buf, "ignored\n{s}\n", .{enc});
    const got = try parseHostDb(std.testing.allocator, contents);
    defer if (got) |g| std.testing.allocator.free(g);
    try std.testing.expectEqualStrings(path, got.?);
}

test "parseMaestralIni reads the sync path" {
    const ini = "[main]\nfoo = bar\n[sync]\npath = /Users/me/Dropbox\n";
    try std.testing.expectEqualStrings("/Users/me/Dropbox", parseMaestralIni(ini).?);
}

test "parseMaestralIni returns null without a sync path" {
    try std.testing.expect(parseMaestralIni("[main]\nfoo = bar\n") == null);
}
